pkg/chunkenc/unordered.go

```diff
@@ -512,3 +517,77 @@ func HeadFromCheckpoint(b []byte, desired HeadBlockFmt) (HeadBlock, error) {
 	return decodedBlock, nil
 
 }
+
+// Orderable can accumulate over a set of Bounded and
+// determine if they're ordered without overlap
+// Not intended for reuse.
+type Orderable struct {
+	overlap bool
+	maxt    int64
+	// position is chosend via the from ts,
+	// so we must cache the through ts for O(1) lookups
+	xs []Bounded
+}
+
+func (o *Orderable) Bounds() (int64, int64) {
+	if o.Empty() {
+		return 0, 0
+	}
+	mint, _ := o.xs[0].Bounds()
+	return mint, o.maxt
+}
+
+func (o *Orderable) Empty() bool { return len(o.xs) == 0 }
+
+func (o *Orderable) Append(b Bounded) {
```

Comment:
This function is scary !

We're going to search, append to a nil array,make a new array, copy an array....

I would prefer a logic where first we check if there's overlap, if not then the normal code is used otherwise we use this `Orderable`.

I honestly think we don't need this `Orderable`, we could simply just sort and verify the overlap.

Some things along those lines:

```
type BoundedIterator interface {
  Bounded
  Iterator() it.EntryIterator
}
```
Should be enough to just order correctly iterators.

```
func IsOverlapping(in []BoundedIterator) bool {
 ....
}

func Order(in []BoundedIterator)  {
         slice.Sort(....)
}
```

I think we should avoid taking a hit for all chunks, while only the one in ingesters might need this.

Reply:
Yeah, thanks for the detailed write up. I'll refactor this.